src/app/qgisapp.cpp

```diff
@@ -9588,6 +9589,47 @@ void QgisApp::pasteFromClipboard( QgsMapLayer *destinationLayer )
   // now create new feature using pasted feature as a template. This automatically handles default
   // values and field constraints
   QgsFeatureList newFeatures {QgsVectorLayerUtils::createFeatures( pasteVectorLayer, newFeaturesDataList, &context )};
+
+  // check constraints
+  QgsFeatureList validFeatures = newFeatures;
+  QgsFeatureList invalidFeatures;
+  for ( const QgsFeature &f : qgis::as_const( validFeatures ) )
```

Comment:
Could we skip this whole loop if no constraints are present? I'm concerned about a regression in paste time for huge numbers of features.